### PR TITLE
backport: bitcoin#15141, #15437, #16240, #18923, #19219, #19277, #20016, #20671 (auxiliary backports)

### PR DESCRIPTION
**Pull request by kittywhiskers**
### Depends on

* https://github.com/dashpay/dash/pull/4885
* https://github.com/dashpay/dash/pull/4891
* https://github.com/dashpay/dash/pull/4892 (optional)

### CODE REVIEW

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
pls see below + d7784a55ec49db7a799447c2986c4ad7878a9480

Also,
- e8d71fff8830103aae3625bfc86ae900f8f884ce and b94edd75b5a40f22252f91b8492bd3546cb6e521 are not backports and looks unrelated, should be in their own PRs imo
- 15141 and 19219 backports are large enough and pretty critical, I would prefer them in a separate PR/PRs
- not sure about 15437, I remember mobile teams were against bip61 removal https://github.com/dashpay/dash/pull/3414#discussion_r409574864. Pinging @QuantumExplorer @HashEngineering.

**Comment by kittywhiskers** [comment]
* This pull request was made with https://github.com/dashpay/dash/pull/4829 in mind and so, attempted to meet all the dependencies that it needed.
  * I use the "auxiliary backports" tag as an opportunity to do miscellaneous, minor, unrelated and/or loose end (i.e. completion of partial) backports, that's why we have this trail mix, it's in the name:tm:.
* https://github.com/dashpay/dash/commit/e8d71fff8830103aae3625bfc86ae900f8f884ce was done here so that I can keep https://github.com/dashpay/dash/pull/4829 clean as it's already quite cluttered
* https://github.com/dashpay/dash/commit/b94edd75b5a40f22252f91b8492bd3546cb6e521 was done so that I can complete https://github.com/dashpay/dash/pull/4830/commits/5cb63f853e0fa063db94fa4fa2a02fb398f10afa (migration to `Check()` meant I was forced to bifurcate what is otherwise, essentially two distinct RPC functions) 
* Not opposed to splitting 15141 and 19219 backports, sounds reasonable to me. (pinging @PastaPastaPasta, thoughts? objections?)
* For 15437, am waiting on mobile team to give their opinion.

File: test/functional/wallet_dump.py [hunk 1]
```diff
@@ -152,5 +152,10 @@ def run_test(self):
         result = self.nodes[0].getaddressinfo(multisig_addr)
         assert result['ismine'] == True
 
+        # self.log.info('Check that wallet is flushed')
+        # with self.nodes[0].assert_debug_log(['Flushing wallet.dat'], timeout=20):
+        #     self.nodes[0].getnewaddress()
+        #
+
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply test/functional/wallet_dump.py:159]: * In my defence, I tried 
  * Reverting the location of the `args.GetBoolArg` check
  * Reverting the PR itself and adding only the test to see if it's a regression (this was a belt-and-suspenders check if I'm not wrong)
  * Removing the GetBoolArg check entirely and letting `MaybeCompactWalletDB` always run its course
 
   None of the three worked and this seemed like the least worst way to go about it, at the time. Will un-comment code and cherry-pick fix, thanks!

**Comment by UdjinM6** [comment]
ok, I see. pls see 1d9d1b7aafe8667b2d8b2540d489827c92a6b4bb for more fixes

**Comment by kittywhiskers** [comment]
Cherry-picked, squashed and updated!

File: src/rpc/client.cpp [hunk 2]
```diff
@@ -165,6 +165,7 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "logging", 0, "include" },
     { "logging", 1, "exclude" },
     { "spork", 1, "value" },
+    { "sporkupdate", 1, "value" },
```
Comment (PastaPastaPasta) [inline src/rpc/client.cpp:168]: b94edd75b5a40f22252f91b8492bd3546cb6e521 needs a release notes, and really probably should be in it's own PR

File: doc/bips.md [hunk 3]
```diff
@@ -15,7 +15,7 @@ BIPs that are implemented by Dash Core (up-to-date up to **v18.0**):
...
 * [`BIP 37`](https://github.com/bitcoin/bips/blob/master/bip-0037.mediawiki): The bloom filtering for transaction relaying, partial Merkle trees for blocks, and the protocol version bump to 70001 (enabling low-bandwidth SPV clients) has been implemented since **v0.8.0** ([PR #1795](https://github.com/bitcoin/bitcoin/pull/1795)).
 * [`BIP 42`](https://github.com/bitcoin/bips/blob/master/bip-0042.mediawiki): The bug that would have caused the subsidy schedule to resume after block 13440000 was fixed in **v0.9.2** ([PR #3842](https://github.com/bitcoin/bitcoin/pull/3842)).
-* [`BIP 61`](https://github.com/bitcoin/bips/blob/master/bip-0061.mediawiki): The 'reject' protocol message (and the protocol version bump to 70002) was added in **v0.9.0** ([PR #3185](https://github.com/bitcoin/bitcoin/pull/3185)). Starting *v0.16.0*, whether to send reject messages can be configured with the `-enablebip61` option.
```
[the remark this answers is not in the document]
Comment (HashEngineering) [reply doc/bips.md:18]: If BIP61 is removed completely, the mobile clients will need a means of detecting errors in a transaction based on the fact that a transaction never is mined into the blockchain.

Currently, the mobile clients depend on BIP61, so I would suggest that before BIP61 support is removed the mobile team makes changes to get around this missing feature to detect errors.

**Comment by kittywhiskers** [comment]
Based on @UdjinM6's suggestion, backports of 15141, and 19219 have been split off to https://github.com/dashpay/dash/pull/4891 and BIP61 removal is now split off to https://github.com/dashpay/dash/pull/4892 so that the PR can proceed while mobile team can figure things out. Is that okay, @HashEngineering?

File: doc/bips.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply doc/bips.md]: Moved BIP61 removal to https://github.com/dashpay/dash/pull/4892

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by PastaPastaPasta — DISMISSED** [review]
utACK for merging via merge commit

**Comment by PastaPastaPasta** [comment]
IMO let's merge https://github.com/dashpay/dash/pull/4880/commits first @UdjinM6

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**